src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubRDDIterator.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.plugin.gcp.publisher.source;
+
+import com.google.api.gax.core.FixedCredentialsProvider;
+import com.google.auth.Credentials;
+import com.google.cloud.pubsub.v1.stub.GrpcSubscriberStub;
+import com.google.cloud.pubsub.v1.stub.SubscriberStub;
+import com.google.cloud.pubsub.v1.stub.SubscriberStubSettings;
+import com.google.pubsub.v1.AcknowledgeRequest;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.PullRequest;
+import com.google.pubsub.v1.PullResponse;
+import com.google.pubsub.v1.ReceivedMessage;
+import org.apache.spark.TaskContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.collection.Iterator;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Iterator for PubSub RDD.
+ * Fetches a fixed amount of messages at a time and acknowledges them.
+ * Finishes when the acknowledged messages are completed and the batch time limit is met.
+ * Returns immediately if there are no more messages to read.
+ */
+public class PubSubRDDIterator implements Iterator<PubSubMessage> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PubSubRDDIterator.class);
+  private static final int MAX_MESSAGES = 1000;
+
+  private final long startTime;
+  private final PubSubSubscriberConfig config;
+  private final TaskContext context;
+  private final long batchDuration;
+  private final Credentials credentials;
+  private final String subscriptionFormatted;
+  private PullRequest pullRequest;
+  private SubscriberStub subscriber;
+  private boolean autoAcknowledge;
+  private Queue<ReceivedMessage> receivedMessages;
+  private List<String> ackIds;
+  private long messageCount;
+
+  public PubSubRDDIterator(PubSubSubscriberConfig config, TaskContext context, long batchDuration,
+                           boolean autoAcknowledge, Credentials credentials) {
+    this.config = config;
+    this.context = context;
+    this.batchDuration = batchDuration;
+    this.credentials = credentials;
+    this.startTime = System.currentTimeMillis();
```

Comment:
shouldn't this be the Time from InputDStream's `compute(Time validTime)` method?

Reply:
DStream's compute time is the time when the job is generated. The actual time when the RDD is reading data from Pub/Sub may be later depending on the processing delays for each batch. Current logic reads for the whole batch duration regardless of when the batch was scheduled. If we consider batch time instead, some batches may not be able to read data.

Reply:
Isn't it better not to read data in those situations though? It would basically be doing backpressure.

The concern is if it consistently takes longer to process the microbatch, the jobs will keep piling up. For example, if there are more partitions than executors available to process them in parallel, we can run into this situation.

Reply:
Makes sense. Made the change to use batch time.